Content.Server/Anomaly/Effects/ReagentProducerAnomalySystem.cs

```diff
@@ -0,0 +1,199 @@
+using System.Linq;
+using Content.Server.Anomaly.Components;
+using Content.Server.Chemistry.EntitySystems;
+using Content.Shared.Anomaly.Components;
+using Robust.Shared.Random;
+using Content.Shared.Chemistry.Reagent;
+using Content.Shared.Chemistry.Components;
+using Robust.Shared.Prototypes;
+using Content.Shared.Sprite;
+using Robust.Server.GameObjects;
+
+namespace Content.Server.Anomaly.Effects;
+
+/// <see cref="ReagentProducerAnomalyComponent"/>
+
+public sealed class ReagentProducerAnomalySystem : EntitySystem
+{
+    //The idea is to divide substances into several categories.
+    //The anomaly will choose one of the categories with a given chance based on severity.
+    //Then a random substance will be selected from the selected category.
+    //There are the following categories:
+
+    //Dangerous:
+    //selected most often. A list of substances that are extremely unpleasant for injection.
+
+    //Fun:
+    //Funny things have an increased chance of appearing in an anomaly.
+
+    //Useful:
+    //Those reagents that the players are hunting for. Very low percentage of loss.
+
+    //Other:
+    //All reagents that exist in the game, with the exception of those prescribed in other lists and the blacklist.
+    //They have low chances of appearing due to the fact that most things are boring and do not bring a
+    //significant effect on the game.
+
+    [Dependency] private readonly SolutionContainerSystem _solutionContainer = default!;
+    [Dependency] private readonly IRobustRandom _random = default!;
+    [Dependency] private readonly IPrototypeManager _proto = default!;
+    [Dependency] private readonly AnomalySystem _anomaly = default!;
+    [Dependency] private readonly PointLightSystem _light = default!;
+    [Dependency] private readonly IPrototypeManager _prototypeManager = default!;
+    [Dependency] private readonly SharedAudioSystem _audio = default!;
+    public override void Initialize()
+    {
+        SubscribeLocalEvent<ReagentProducerAnomalyComponent, AnomalySeverityChangedEvent>(OnSeverityChanged);
+        SubscribeLocalEvent<ReagentProducerAnomalyComponent, AnomalyPulseEvent>(OnPulse);
+        SubscribeLocalEvent<ReagentProducerAnomalyComponent, MapInitEvent>(OnMapInit);
+    }
+
+    private void OnPulse(EntityUid uid, ReagentProducerAnomalyComponent component, ref AnomalyPulseEvent args)
+    {
+        ChangeReagent(uid, component, args.Severity, args.Stability);
+    }
+
+    private void ChangeReagent(EntityUid uid, ReagentProducerAnomalyComponent component, float severity, float stability )
+    {
+        if (_random.NextFloat(0.0f, 1.0f) > stability)
+        {
+            var reagent = GetRandomReagentType(uid, component, severity);
+            component.ProducingReagent = reagent;
+            _audio.PlayPvs(component.ChangeSound, uid);
+        }
+    }
```

Comment:
it makes more sense to me that `ChangeReagent` shouldnt do any probability checks, and should just always change the reagent--`OnPulse` should do the `_random.NextFloat(0.0f, 1.0f) > stability` check itself

Reply:
Yes, it's logical. I don't know why I did it so strangely 🤔